src/exec/use_pty/pipe.rs

```diff
@@ -142,78 +156,202 @@ impl<R: Read, W: Write> Buffer<R, W> {
+            // head.
+            //     ┌───────────┐
+            // ┌───┬───┬───┬───┬───┬───┬───┬───┐
+            // │ D │   │   │   │ M │ D │ D │ D │
+            // └───┴───┴───┴───┴───┴───┴───┴───┘
+            //       ▲               ▲
+            //       │               │
+            //     tail             head
+            //
+            // In the extreme case where tail is as close as it can be to head, the sub-buffer will
+            // have a length of one. We can be sure that the marker byte is not at tail because the
+            // buffer is not full.
+            //     ┌───┐
+            // ┌───┬───┬───┬───┬───┬───┬───┬───┐
+            // │ D │   │ M │ D │ D │ D │ D │ D │
+            // └───┴───┴───┴───┴───┴───┴───┴───┘
+            //       ▲       ▲
+            //       │       │
+            //     tail     head
+            &mut self.buffer[self.tail..(self.head - 1)]
+        };
+
+        // Insert `len` bytes from `read` into the buffer.
         let len = read.read(buffer)?;
 
-        // Mark the `len` bytes after the busy section as busy too.
-        self.end += len;
-
-        // If we read something, the buffer is not empty anymore and we can resume writing.
         if len > 0 {
+            // We update tail so it becomes the next position where we can insert data again.
+            self.tail += len;
+            // However, we must be sure that tail doesn't go past the last position.
+            self.tail %= self.buffer.len();
+            // If we actually inserted something, the buffer is not empty anymore and we can resume
+            // writing.
             self.write_handle.resume(registry);
         }
 
         Ok(())
     }
 
-    /// Write bytes from the buffer.
+    /// Remove bytes from the buffer and write them.
     ///
     /// Calling this function will block until `write` is ready to be written.
-    fn write<T: Process>(
+    fn remove<T: Process>(
         &mut self,
         write: &mut W,
         registry: &mut EventRegistry<T>,
     ) -> io::Result<()> {
-        // Don't write if the buffer is empty.
+        // Don't remove anything if the buffer is empty.
         if self.is_empty() {
             self.write_handle.ignore(registry);
             return Ok(());
         }
 
-        // This is the busy section of the buffer.
-        let buffer = &self.buffer[self.start..self.end];
-
-        // Write the first `len` bytes of the busy section to `write`.
-        let len = write.write(buffer)?;
-
-        if len == buffer.len() {
-            // If we were able to write all the busy section, we can mark the whole buffer as free.
-            self.start = 0;
-            self.end = 0;
+        // We need to find the sub-buffer with data that starts at head. To do this, we must ensure
+        // that we're not including the end byte or any other empty byte after it.
+        let buffer = if self.head < self.tail {
```

Comment:
shouldn't this be `head <= tail`? when `head == tail` the ring buffer is empty so `buffer` should be an empty slice. as currently written, this will return uninitialized / old data on `head == tail` because the `else` branch is taken

Reply:
`head` cannot be equal to `tail` because this function checks earlier if the buffer is empty and returns preemptively.